### PR TITLE
[AI] Refactor shipyards using BuildingType

### PR DESCRIPTION
**Pull request by Grummel7**
This is mostly refactoring: Disentangle shipyards from the hulk of spaghetti currently producing building. Also changed some names to improve readability and got rid of hard-coded building costs.
By using BuildingType and _try_enqueue, some code became a lot shorter, too, and as a side-effect the AI no longer completely ignores dislikes. Kobuntera are so much more productive when they do not have to build ships.
Last but not least, I fixed an old bug that prevented the AI from building Solar Containment Units.

Some of the extracted functions could do with some further refactoring, but one step at a time...

### CODE REVIEW

File: default/python/AI/AIDependencies.py
Comment (Grummel7) [inline default/python/AI/AIDependencies.py:813]: In the long term, code from building_locations.py should be completely moved to buildings.py, then this set can also be moved there (and become a set or enums).

**Comment by Cjkjvfnby** [comment]
> Some of the extracted functions could do with some further refactoring, but one step at a time...

It is not a one step, it's a list of them. :)

File: default/python/AI/ProductionAI.py
Comment (Cjkjvfnby) [inline default/python/AI/ProductionAI.py]: It's a good practice to declare a variable as close to its use as possible.  It makes code easier to follow.

File: default/python/AI/ProductionAI.py
Comment (Cjkjvfnby) [inline default/python/AI/ProductionAI.py]: You could use immutable default here.

```suggestion
def _build_ship_facilities(building_type: BuildingType, top_pids: Set[PlanetId] = frozenset()) -> None:
 ```

File: default/python/AI/ProductionAI.py
Comment (Cjkjvfnby) [inline default/python/AI/ProductionAI.py]: You could use set comprehansions

```suggestion
        try_pids = {pid for sys_id in try_systems for pid in get_owned_planets_in_system(sys_id)}
```

File: default/python/AI/ProductionAI.py
Comment (Cjkjvfnby) [inline default/python/AI/ProductionAI.py]: This one is quite long, let's split it into chunks. 

`list(pids for pids in try_pids & top_pids if building_type.can_be_produced(pids))`

Please use list comprehansions.

File: default/python/AI/ProductionAI.py
Comment (Cjkjvfnby) [inline default/python/AI/ProductionAI.py]: `ObjectId` is just typing wrapper around int for better readability, most API does not return it. 

PlanetId not ObjectId

File: default/python/AI/ProductionAI.py
Comment (Cjkjvfnby) [inline default/python/AI/ProductionAI.py:1433]: This change could be done in a separate PR. It's a nonbreaking change so you don't have to wait until the previous PR gets merged.  It was a really good opportunity to parallelize the review process. 

I encourage you to find a catch such cases in the future. 

- This makes PRs small, and smaller pr tend to have fewer comments. 
- You don't need to wait for feature A to be merged because of comments in Feature B.
- Refactoring of logic and refactoring of signatures require different levels of attention, when they are mixed, it draws a bit more reviewer attention to simple cases. 

This PR is already big, so it will take me a couple of passes to review it.  So moving such things to a separate PR will reduce that number.

File: default/python/AI/buildings.py
Comment (Cjkjvfnby) [inline default/python/AI/buildings.py]: This change produces a lot of noise to PR.  Please create a separate one with it. 

These kinds of changes have a high chance to be merged without comment on the same day.

File: default/python/AI/ProductionAI.py
Comment (Cjkjvfnby) [inline default/python/AI/ProductionAI.py:1702]: Stopped here.

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]
Reviewed a part of changes, and left a couple of comments.

File: default/python/AI/ProductionAI.py
Comment (Cjkjvfnby) [inline default/python/AI/ProductionAI.py]: I'd say let's make NamedTuple here. Return structure is quite complicated, and a named tuple will document it well.

File: default/python/AI/ProductionAI.py
Comment (Cjkjvfnby) [inline default/python/AI/ProductionAI.py:1749]: You could extract conditions to the function, and use `key` for sorting.

```suggestion
                def pids_filter(planet_):
                    return (
                            planet_ and
                            planet_.initialMeterValue(fo.meterType.targetPopulation) >= 3.0 and
                            building_type.can_be_produced(planet_.id
                          )
    
                    def sort_key(planet_):
                    return planet_.initialMeterValue(fo.meterType.population)
    
                pids = sorted(
                    (planet_.id for planet_ in planets if pids_filter(planet_)), key=sort_key)
```

File: default/python/AI/ProductionAI.py
Comment (Cjkjvfnby) [inline default/python/AI/ProductionAI.py:1754]: That's a huge indent.   At this point a already forget what was this function name.

Try to run a complexity inspection on this file.
```shell
pip install radon
radon cc AI/ProductionAI.py -s --min C
```

10 is considered normal.

изображение

File: default/python/AI/ProductionAI.py
Comment (Cjkjvfnby) [inline default/python/AI/ProductionAI.py]: Looks like this is unused.

File: default/python/AI/AIDependencies.py
Comment (Cjkjvfnby) [reply default/python/AI/AIDependencies.py:813]: I'd prefer to convert buildings to a module and move this to a separate file.

**Review by Cjkjvfnby — COMMENTED** [review]
Left a couple of comments.

File: default/python/AI/AIDependencies.py
Comment (o01eg) [reply default/python/AI/AIDependencies.py:813]: Are all those buildings have ship hulls or ship parts which require those building? Maybe we can generate this list from FOCS data.

Comment (Grummel7) [reply default/python/AI/AIDependencies.py:813]: > Are all those buildings have ship hulls or ship parts which require those building? Maybe we can generate this list from FOCS data.

I guess we could, but we  still have to adapt the AI if the game gets more ship yards and the AI should learn to use it. I think it is not worth the effort. There'll always be a little code duplication between FOCS and AI.

File: default/python/AI/ProductionAI.py
Comment (Grummel7) [reply default/python/AI/ProductionAI.py]: Yes, I guess I can.
That was not my code, though, my only change in that line was to change the name try_locs to try_pids, because the existing code used _locs for both planets and systems.

File: default/python/AI/ProductionAI.py
Comment (Grummel7) [reply default/python/AI/ProductionAI.py]: Yes, of course it should be PlanetId.
As far as I understood, NewType will actually create a class in python 3.10, so I guess it would be possible to specify it in is_instance as well, wouldn't it?

File: default/python/AI/ProductionAI.py
Comment (Grummel7) [reply default/python/AI/ProductionAI.py:1433]: The problem was, that this extension occurred to me in the middle of all those changes, after writing  _try_enqueue(building_type, [pid]) several times.

My next PR should be parallel to this.

File: default/python/AI/ProductionAI.py
Comment (Grummel7) [reply default/python/AI/ProductionAI.py]: good idea, safes the None-handling

File: default/python/AI/ProductionAI.py
Comment (Cjkjvfnby) [reply default/python/AI/ProductionAI.py:1433]: It's not a problem.  It was before you submitted this PR. 

Here are simple git commands that could help you with it.

- git stash
- git checkout <new_branch>
- git add ...
- git commit
- git push 
- git checkout -
- git stash pop

**Review by Cjkjvfnby — APPROVED** [review]

File: default/python/AI/AIDependencies.py
Comment (Cjkjvfnby) [reply default/python/AI/AIDependencies.py:813]: @o01eg any input from your side?  Otherwise, I am ready to merge.

**Pull request merged**
